src/IlliniBlockchainSP22Token.sol

```diff
@@ -2,47 +2,113 @@
+        _;
     }
 
-    function uri(uint256)
+    function changeOwner(address _owner) public onlyOwner {
+        owner = _owner;
+    }
+
+    function contractURI() public view returns (string memory) {
+        string memory image = Base64.encode(
+            bytes(abi.encodePacked(NFTSVG.baseSVG, "</svg>"))
+        );
+        return
+            string(
+                abi.encodePacked(
+                    "data:application/json;base64,",
+                    Base64.encode(
+                        bytes(
+                            abi.encodePacked(
+                                '{"name":"',
+                                name,
+                                '",',
+                                '"description":"Student organization at the University of Illinois at Urbana-Champaign.",',
+                                '"external_link": "https://illiniblockchain.com/",',
+                                '"image": "',
+                                "data:image/svg+xml;base64,",
+                                image,
+                                '"}'
+                            )
+                        )
+                    )
+                )
+            );
+    }
+
+    function uri(uint256 _tokenId)
         public
-        pure
+        view
         virtual
         override
         returns (string memory)
-    {}
+    {
+        TokenMetadata memory meta = tokenMetadata[_tokenId];
+        string memory yearStr = meta.year.toString();
+        string memory termStr = terms[meta.termId];
+        string memory image = Base64.encode(
+            bytes(
+                NFTSVG.generateSVG(
+                    NFTSVG.SVGParams({year: yearStr, term: termStr})
+                )
+            )
+        );
+        return
+            string(
+                abi.encodePacked(
+                    "data:application/json;base64,",
+                    Base64.encode(
+                        bytes(
+                            abi.encodePacked(
+                                '{"name":"IlliniBlockchain ',
+                                termStr,
+                                " ",
+                                yearStr,
+                                ' Member",',
+                                '"description":"IlliniBlockchain Membership",',
+                                '"image": "',
+                                "data:image/svg+xml;base64,",
+                                image,
+                                '"}'
+                            )
+                        )
+                    )
+                )
+            );
+    }
 
     function mint(
-        uint8 _id,
+        address _to,
+        uint256 _id,
         uint256 _amount,
+        TokenMetadata calldata _metadata,
         bytes memory _data
     ) public onlyOwner {
-        _mint(msg.sender, _id, _amount, _data);
+        tokenMetadata[_id] = _metadata;
```

Comment:
Can previous IDs be overwritten this way? Do we need some kind of tokenID validation?

Reply:
Yeah, good point. Maybe there should just be a different function for setting the metadata. I guess I was treating this more as an ERC 721 contract and forgot that you can still mint more of a specific token id. We can have a `setTokenURI` function that can only be called once for a token.

Reply:
perfect, that's exactly what i was thinking actually -- at first i thought it's cumbersome to set URI, then mint -- but actually, it makes more sense than feeding the URI data on each mint.